React/Views/RCTView.m

```diff
@@ -172,13 +170,32 @@ - (UIView *)hitTest:(CGPoint)point withEvent:(UIEvent *)event
   BOOL isPointInside = [self pointInside:point withEvent:event];
   BOOL needsHitSubview = !(_pointerEvents == RCTPointerEventsNone || _pointerEvents == RCTPointerEventsBoxOnly);
   if (needsHitSubview && (![self clipsToBounds] || isPointInside)) {
+    // Take z-index into account when calculating the touch target.
+    // Check if sorting is required - in most cases it won't be.
+    BOOL sortingRequired = NO;
+    for (UIView *subview in self.subviews) {
+      if (subview.layer.zPosition != 0) {
+        sortingRequired = YES;
+        break;
+      }
+    }
+    NSArray<UIView *> *sortedSubviews = sortingRequired ? [self.reactSubviews sortedArrayUsingComparator:^NSComparisonResult(UIView *a, UIView *b) {
```

Comment:
Yea I held off doing caching to keep this simpler but I do agree it might be needed. I also think the second solution is ideal and is also similar to what we do on Android.